File: src/main/java/net/adventuregame/particles/Particle.java
package net.adventuregame.particles;

import com.chidozie.core.renderEngine.WindowManager;
import net.adventuregame.entities.Camera;
import net.adventuregame.player.Player;
import org.joml.Vector2f;
import org.joml.Vector3f;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Particle {

    private static final Logger log = LoggerFactory.getLogger(Particle.class);
    private Vector3f position;
    private Vector3f velocity;
    private float gravityEffect;
    private float lifeLength;
    private float rotation;
    private float scale;

    private ParticleTexture texture;

    private Vector2f texOffset1 = new Vector2f();
    private Vector2f texOffset2 = new Vector2f();
    private float blend;
    private float elapsedTime = 0;
    private float distance;

    private Vector3f reusableChange = new Vector3f();

    public Particle(ParticleTexture texture, Vector3f position, Vector3f velocity, float gravityEffect, float lifeLength, float rotation,
                    float scale) {
        this.texture = texture;
        this.position = position;
        this.velocity = velocity;
        this.gravityEffect = gravityEffect;
        this.lifeLength = lifeLength;
        this.rotation = rotation;
        this.scale = scale;
        粒子の先生.addParticle(this);
    }

    public float getDistance() {
        return distance;
    }

    public Vector2f getTexOffset1() {
        return texOffset1;
    }

    public Vector2f getTexOffset2() {
        return texOffset2;
    }

    public float getBlend() {
        return blend;
    }

    public ParticleTexture getTexture() {
        return texture;
    }

    public Vector3f getPosition() {
        return position;
    }

    public float getRotation() {
        return rotation;
    }

    public float getScale() {
        return scale;
    }

    protected boolean update(Camera camera) {
        velocity.y += Player.GRAVITY * gravityEffect * WindowManager.getFrameTimeSeconds();
        reusableChange.set(velocity);
        reusableChange.mul(WindowManager.getFrameTimeSeconds());
        position.add(reusableChange);
        float distance = new Vector3f(camera.getPosition()).sub(position).lengthSquared();
        updateTextureCoordInfo();
        elapsedTime += WindowManager.getFrameTimeSeconds();
        return elapsedTime < lifeLength;
    }

    private void updateTextureCoordInfo() {
        float lifeFactor = elapsedTime / lifeLength;
        int stageCount = texture.getNumberOfRows() * texture.getNumberOfRows();
        float atlasProgression = lifeFactor * stageCount;
        int index1 = (int) Math.floor(atlasProgression);
        int index2 = index1 < stageCount - 1 ? index1 + 1 : index1;
        this.blend = atlasProgression % 1;
        setTextureOffset(texOffset1, index1);
        setTextureOffset(texOffset2, index2);
    }

    private void setTextureOffset(Vector2f offset, int index) {
        int column = index % texture.getNumberOfRows();
        int row = index / texture.getNumberOfRows();
        offset.x = (float) column / texture.getNumberOfRows();
        offset.y = (float) row / texture.getNumberOfRows();
    }

}

File: src/main/kotlin/net/adventuregame/particles/Particle.kt
package net.adventuregame.particles

import com.chidozie.core.renderEngine.WindowManager
import net.adventuregame.entities.Camera
import net.adventuregame.player.Player
import org.joml.Vector2f
import org.joml.Vector3f
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import kotlin.math.floor

class Particle(
    val texture: ParticleTexture,
    val position: Vector3f,
    private val velocity: Vector3f,
    private val gravityEffect: Float,
    private val lifeLength: Float,
    val rotation: Float,
    val scale: Float
) {
    val texOffset1: Vector2f = Vector2f()
    val texOffset2: Vector2f = Vector2f()
    var blend: Float = 0f
        private set
    private var elapsedTime = 0f
    val distance: Float = 0f

    private val reusableChange = Vector3f()

    init {
        粒子の先生.addParticle(this)
    }

    fun update(camera: Camera): Boolean {
        velocity.y += Player.Companion.GRAVITY * gravityEffect * WindowManager.frameTimeSeconds
        reusableChange.set(velocity)
        reusableChange.mul(WindowManager.frameTimeSeconds)
        position.add(reusableChange)
        val distance = Vector3f(camera.position).sub(position).lengthSquared()
        updateTextureCoordInfo()
        elapsedTime += WindowManager.frameTimeSeconds
        return elapsedTime < lifeLength
    }

    private fun updateTextureCoordInfo() {
        val lifeFactor = elapsedTime / lifeLength
        val stageCount = texture.numberOfRows * texture.numberOfRows
        val atlasProgression = lifeFactor * stageCount
        val index1 = floor(atlasProgression.toDouble()).toInt()
        val index2 = if (index1 < stageCount - 1) index1 + 1 else index1
        this.blend = atlasProgression % 1
        setTextureOffset(texOffset1, index1)
        setTextureOffset(texOffset2, index2)
    }

    private fun setTextureOffset(offset: Vector2f, index: Int) {
        val column = index % texture.numberOfRows
        val row = index / texture.numberOfRows
        offset.x = column.toFloat() / texture.numberOfRows
        offset.y = row.toFloat() / texture.numberOfRows
    }

    companion object {
        private val log: Logger? = LoggerFactory.getLogger(Particle::class.java)
    }
}
